platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/proof/algorithms/StateProofValidator.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.merkle.proof.algorithms;
+
+import static com.swirlds.common.crypto.DigestType.SHA_384;
+
+import com.swirlds.common.crypto.Cryptography;
+import com.swirlds.common.merkle.proof.tree.StateProofInternalNode;
+import com.swirlds.common.merkle.proof.tree.StateProofNode;
+import com.swirlds.common.system.NodeId;
+import com.swirlds.common.system.address.Address;
+import com.swirlds.common.system.address.AddressBook;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.security.MessageDigest;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Utility methods for validating state proofs.
+ */
+public final class StateProofValidator {
+
+    private StateProofValidator() {}
+
+    public static long computeValidSignatureWeight(
+            @NonNull final AddressBook addressBook,
+            @NonNull final List<NodeSignature> signatures,
+            @NonNull final SignatureValidator signatureValidator,
+            @NonNull final byte[] hashBytes) {
+
+        final Set<NodeId> signingNodes = new HashSet<>();
+
+        long validWeight = 0;
+        for (final NodeSignature nodeSignature : signatures) {
+            if (!signingNodes.add(nodeSignature.nodeId())) {
+                // Signature is not unique.
+                continue;
+            }
+
+            if (!addressBook.contains(nodeSignature.nodeId())) {
+                // Signature is not in the address book.
+                continue;
+            }
+            final Address address = addressBook.getAddress(nodeSignature.nodeId());
+            if (address.getWeight() == 0) {
+                // Don't bother validating the signature of a zero weight node.
+                continue;
+            }
+
+            // nodeSignature.signature.verifySignature(hashBytes, address.getSigPublicKey())
+            if (!signatureValidator.verifySignature(nodeSignature.signature(), hashBytes, address.getSigPublicKey())) {
```

Comment:
do you want this to throw an exception if this public key is null?

If not, you should add a null check to this condition